File: 프로그래머스/lv3/43162. 네트워크/네트워크.py

def union(a,b,l):
    pa = find(a,l)
    pb = find(b,l)
    
    if(pa<=pb):
        l[pb] = pa
    else:
        l[pa] = pb

def find(a,p):
    r = p[a]
    
    while(p[r]!=r):
        r = p[r]
        
    
    tmp = a
    
    while(tmp!=p[tmp]):
        pa = p[tmp]
        p[tmp] = r
        tmp = pa
    
    return r
        
def solution(n, computers):
    l = [i for i in range(n+1)]
    
    for i in range(n):
        for j in range(n):
            if i==j:
                continue
            if(computers[i][j] == 1):
                union(i+1,j+1,l)
    
        
        for i in range(n):
            l[i] = find(i,l)
    
    return len(set(l[1:])) 

File: 프로그래머스/lv3/43162. 네트워크/네트워크.swift
import Foundation

func find(_ a:Int,_ root: inout [Int]) -> Int {
    
    var r = root[a]
    
    while( r != root[r] ){
        r = root[r]
    }
    
    var c = a
    
    while( c != root[c] ) {
        
        let pc = root[c]
        root[c] = r
        c = pc 
    }
    
    return r
    
}

func union(_ a:Int,_ b:Int,_ root: inout [Int]) {
 
    let ra = find(a,&root)
    let rb = find(b,&root)
    
    if ra <= rb {
        root[rb] = ra
    }
    else {
        root[ra] = rb 
    }
    
}


func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    
    var root:[Int] = [Int](0..<n)
    
    for i in (0..<n) {
        for j in (0..<n){
            
            if i == j { continue }
            
            if computers[i][j] == 1 && find(i,&root) != find(j,&root) {
                union(i,j,&root)
            } 
            
        }
        
        for k in 0..<n {
            root[k] = find(k,&root)
        }
        
    }
    
    return Set(root).count
}
